rust/src/lib.rs

```diff
@@ -2932,6 +2932,252 @@ pub unsafe extern "C" fn zcashlc_create_proposed_transactions(
+///   function call.
+/// - The total size `spend_params_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
+/// - `output_params` must be non-null and valid for reads for `output_params_len` bytes,
+///   and it must have an alignment of `1`. Its contents must be the Sapling output
+///   proving parameters.
+/// - The memory referenced by `output_params` must not be mutated for the duration of the
+///   function call.
+/// - The total size `output_params_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of pointer::offset.
+/// - Call [`zcashlc_free_boxed_slice`] to free the memory associated with the returned
+///   pointer when done using it.
+#[no_mangle]
+pub extern "C" fn zcashlc_add_proofs_to_pczt(
+    pczt_ptr: *const u8,
+    pczt_len: usize,
+    spend_params: *const u8,
+    spend_params_len: usize,
+    output_params: *const u8,
+    output_params_len: usize,
+) -> *mut FfiBoxedSlice {
+    let res = catch_panic(|| {
+        let pczt_bytes = unsafe { slice::from_raw_parts(pczt_ptr, pczt_len) };
+        let pczt = Pczt::parse(pczt_bytes).map_err(|e| anyhow!("Invalid PCZT: {:?}", e))?;
+        let spend_params = Path::new(OsStr::from_bytes(unsafe {
+            slice::from_raw_parts(spend_params, spend_params_len)
+        }));
+        let output_params = Path::new(OsStr::from_bytes(unsafe {
+            slice::from_raw_parts(output_params, output_params_len)
+        }));
+
+        let prover = LocalTxProver::new(spend_params, output_params);
+
+        let pczt_with_proofs = Prover::new(pczt)
+            .create_orchard_proof(&orchard::circuit::ProvingKey::build())
+            .map_err(|e| anyhow!("Failed to create Orchard proof for PCZT: {:?}", e))?
+            .create_sapling_proofs(&prover, &prover)
+            .map_err(|e| anyhow!("Failed to create Sapling proofs for PCZT: {:?}", e))?
+            .finish();
+
+        Ok(FfiBoxedSlice::some(pczt_with_proofs.serialize()))
+    });
+    unwrap_exc_or_null(res)
+}
+
+/// Takes a PCZT that has been separately proven and signed, finalizes it, and stores it
+/// in the wallet.
+///
+/// Returns the txid of the completed transaction as a byte array.
+///
+/// # Safety
+///
+/// - `db_data` must be non-null and valid for reads for `db_data_len` bytes, and it must
+///   have an alignment of `1`. Its contents must be a string representing a valid system
+///   path in the operating system's preferred representation.
+/// - The memory referenced by `db_data` must not be mutated for the duration of the
+///   function call.
+/// - The total size `db_data_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
+/// - `pczt_with_proofs_ptr` must be non-null and valid for reads for `pczt_with_proofs_len` bytes, and it
+///   must have an alignment of `1`. Its contents must be an encoded Proposal protobuf.
```

Comment:
> Its contents must be an encoded Proposal protobuf.

I don't think that's a safety requirement?

Reply:
Note that although it's not a safety requirement, we include what the contents are supposed to be in the safety documentation everywhere (see also the path to `db_data` etc.). Duplicating the parameter documentation seems excessive to me.